File: assets/js/hooks/hover_card.js
import { hasFinePointerHover, whenMouse } from "../lib/pointer";

export const HoverCard = {
  mounted() {
    this.openDelay = parseInt(this.el.dataset.openDelay) || 200;
    this.closeDelay = parseInt(this.el.dataset.closeDelay) || 150;

    this.openTimeout = null;
    this.closeTimeout = null;
    this.isOpen = false;
    this._touchOpen = !hasFinePointerHover();

    this.trigger = this.el.querySelector('[data-hover-card-trigger]');
    this.content = this.el.querySelector('[data-hover-card-content]');

    if (!this.trigger || !this.content) return;

    this.onTriggerEnter = whenMouse(this.onTriggerEnter.bind(this));
    this.onTriggerLeave = whenMouse(this.onTriggerLeave.bind(this));
    this.onContentEnter = whenMouse(this.onContentEnter.bind(this));
    this.onContentLeave = whenMouse(this.onContentLeave.bind(this));
    this.onTriggerClick = this.onTriggerClick.bind(this);
    this.onDocumentPointerDown = this.onDocumentPointerDown.bind(this);

    if (this._touchOpen) {
      this.trigger.addEventListener('click', this.onTriggerClick);
      document.addEventListener('pointerdown', this.onDocumentPointerDown, true);
    } else {
      this.trigger.addEventListener('pointerenter', this.onTriggerEnter);
      this.trigger.addEventListener('pointerleave', this.onTriggerLeave);
      this.content.addEventListener('pointerenter', this.onContentEnter);
      this.content.addEventListener('pointerleave', this.onContentLeave);
    }
  },

  onTriggerEnter() {
    this.show();
  },

  onTriggerLeave() {
    this.hide();
  },

  onContentEnter() {
    // Cancel close when hovering over the content
    if (this.closeTimeout) {
      clearTimeout(this.closeTimeout);
      this.closeTimeout = null;
    }
  },

  onContentLeave() {
    this.hide();
  },

  onTriggerClick(e) {
    e.preventDefault();
    e.stopPropagation();
    if (this.isOpen) {
      this.closeNow();
    } else {
      this.show(true);
    }
  },

  onDocumentPointerDown(e) {
    if (!this.isOpen) return;
    if (this.trigger?.contains(e.target) || this.content?.contains(e.target)) {
      return;
    }
    this.closeNow();
  },

  show(immediate = false) {
    if (this.closeTimeout) {
      clearTimeout(this.closeTimeout);
      this.closeTimeout = null;
    }

    if (this.isOpen) return;

    const open = () => {
      this.content.style.display = 'block';
      this.positionContent();
      this.content.dataset.state = 'open';
      this.trigger.dataset.state = 'open';
      this.isOpen = true;
      this.openTimeout = null;
    };

    if (immediate) {
      if (this.openTimeout) {
        clearTimeout(this.openTimeout);
        this.openTimeout = null;
      }
      open();
      return;
    }

    if (this.openTimeout) return;
    this.openTimeout = setTimeout(open, this.openDelay);
  },

  hide() {
    if (this.openTimeout) {
      clearTimeout(this.openTimeout);
      this.openTimeout = null;
    }

    if (!this.closeTimeout && this.isOpen) {
      this.closeTimeout = setTimeout(() => {
        this.closeNow();
      }, this.closeDelay);
    }
  },

  closeNow() {
    if (this.openTimeout) clearTimeout(this.openTimeout);
    if (this.closeTimeout) clearTimeout(this.closeTimeout);
    this.openTimeout = null;
    this.closeTimeout = null;
    if (!this.isOpen) return;
    this.content.dataset.state = 'closed';
    this.content.style.display = 'none';
    this.trigger.dataset.state = 'closed';
    this.isOpen = false;
  },

  positionContent() {
    const triggerEl = this.trigger.firstElementChild || this.trigger;
    const triggerRect = triggerEl.getBoundingClientRect();
    const side = this.content.dataset.side || 'bottom';
    const align = this.content.dataset.align || 'start';
    const sideOffset = 8;
    const collisionPadding = 10;

    // Temporarily make visible for measurement
    this.content.style.position = 'fixed';
    this.content.style.visibility = 'hidden';
    this.content.style.display = 'block';

    const contentWidth = this.content.offsetWidth;
    const contentHeight = this.content.offsetHeight;

    this.content.style.visibility = '';

    let top, left;

    if (side === 'bottom') {
      top = triggerRect.bottom + sideOffset;
    } else if (side === 'top') {
      top = triggerRect.top - contentHeight - sideOffset;
    } else if (side === 'right') {
      left = triggerRect.right + sideOffset;
      top = triggerRect.top;
    } else {
      left = triggerRect.left - contentWidth - sideOffset;
      top = triggerRect.top;
    }

    if (side === 'top' || side === 'bottom') {
      if (align === 'start') {
        left = triggerRect.left;
      } else if (align === 'end') {
        left = triggerRect.right - contentWidth;
      } else {
        left = triggerRect.left + (triggerRect.width - contentWidth) / 2;
      }
    }

    const viewportWidth = window.innerWidth;
    const viewportHeight = window.innerHeight;

    if (left < collisionPadding) left = collisionPadding;
    if (left + contentWidth > viewportWidth - collisionPadding) {
      left = viewportWidth - contentWidth - collisionPadding;
    }

    if (top + contentHeight > viewportHeight - collisionPadding) {
      // Flip to top if needed
      if (side === 'bottom') {
        top = triggerRect.top - contentHeight - sideOffset;
      } else {
        top = viewportHeight - contentHeight - collisionPadding;
      }
    }

    this.content.style.position = 'fixed';
    this.content.style.top = top + 'px';
    this.content.style.left = left + 'px';
    this.content.style.zIndex = '9999';
  },

  destroyed() {
    if (this.openTimeout) clearTimeout(this.openTimeout);
    if (this.closeTimeout) clearTimeout(this.closeTimeout);

    if (this.trigger) {
      this.trigger.removeEventListener('pointerenter', this.onTriggerEnter);
      this.trigger.removeEventListener('pointerleave', this.onTriggerLeave);
      this.trigger.removeEventListener('click', this.onTriggerClick);
    }
    if (this.content) {
      this.content.removeEventListener('pointerenter', this.onContentEnter);
      this.content.removeEventListener('pointerleave', this.onContentLeave);
    }
    document.removeEventListener('pointerdown', this.onDocumentPointerDown, true);
  }
};

File: lib/essence_ui/primitives/hover_card.ex
defmodule EssenceUI.Primitives.HoverCard do
  @moduledoc false

  use EssenceUI.Primitives

  attr :id, :string, required: true
  attr :open, :boolean, default: false
  attr :default_open, :boolean, default: false
  attr :open_delay, :integer, default: 200
  attr :close_delay, :integer, default: 150
  attr :on_open_change, :string, default: nil
  attr :rest, :global
  slot :inner_block, required: true

  def root(assigns) do
    open = assigns.open || assigns.default_open
    assigns = assign(assigns, :state, if(open, do: "open", else: "closed"))

    ~H"""
    <div
      id={@id}
      phx-hook="HoverCardRoot"
      data-state={@state}
      data-open-delay={@open_delay}
      data-close-delay={@close_delay}
      data-on-open-change={@on_open_change}
      {@rest}
    >
      {render_slot(@inner_block)}
    </div>
    """
  end

  attr :id, :string, default: nil
  attr :content_id, :string, default: nil
  attr :rest, :global
  slot :inner_block, required: true

  def trigger(assigns) do
    ~H"""
    <div
      id={@id}
      data-essence-hover-card-trigger
      aria-describedby={@content_id}
      data-state="closed"
      {@rest}
    >
      {render_slot(@inner_block)}
    </div>
    """
  end

  attr :id, :string, required: true
  attr :target, :string, default: "body"
  attr :class, :any, default: nil
  attr :container, :string, default: "div"
  slot :inner_block, required: true

  def portal(assigns) do
    ~H"""
    <Phoenix.Component.portal id={@id} target={@target} class={@class} container={@container}>
      {render_slot(@inner_block)}
    </Phoenix.Component.portal>
    """
  end

  attr :id, :string, default: nil
  attr :side, :string, values: ["top", "right", "bottom", "left"], default: "bottom"
  attr :align, :string, values: ["start", "center", "end"], default: "center"
  attr :side_offset, :integer, default: 8
  attr :rest, :global
  slot :inner_block, required: true

  def content(assigns) do
    ~H"""
    <div
      id={@id}
      data-essence-hover-card-content
      data-state="closed"
      data-side={@side}
      data-align={@align}
      data-side-offset={@side_offset}
      style="display: none; position: fixed; z-index: 50;"
      {@rest}
    >
      {render_slot(@inner_block)}
    </div>
    """
  end

  attr :id, :string, default: nil
  attr :rest, :global
  slot :inner_block

  def arrow(assigns) do
    ~H"""
    <span id={@id} data-essence-hover-card-arrow aria-hidden="true" {@rest}>
      {render_slot(@inner_block)}
    </span>
    """
  end
end
